sys/aarch64/board.c

```diff
@@ -76,16 +119,31 @@ static void rpi3_physmem(void) {
   paddr_t kern_end = (paddr_t)_bootmem_end;
   paddr_t rd_start = ramdisk_get_start();
   paddr_t rd_end = rd_start + ramdisk_get_size();
+  paddr_t dtb_start = dtb_early_root();
+  paddr_t dtb_end = dtb_start + fdt_totalsize(PHYS_TO_DMAP(dtb_start));
+
+  addr_range_t memory[3] = {
+    {.start = rounddown(kern_start, PAGESIZE),
+     .end = roundup(kern_end, PAGESIZE)},
+    {.start = rounddown(rd_start, PAGESIZE), .end = roundup(rd_end, PAGESIZE)},
+    {.start = rounddown(dtb_start, PAGESIZE),
+     .end = roundup(dtb_end, PAGESIZE)},
+  };
 
-  vm_physseg_plug(ram_start, kern_start);
-  vm_physseg_plug_used(kern_start, kern_end);
+  qsort(memory, 3, sizeof(addr_range_t), addr_range_cmp);
 
-  if (rd_start != rd_end) {
-    vm_physseg_plug(kern_end, rd_start);
-    vm_physseg_plug_used(rd_start, rd_end);
-    vm_physseg_plug(rd_end, ram_end);
-  } else {
-    vm_physseg_plug(kern_end, ram_end);
+  vm_physseg_plug(ram_start, memory[0].start);
+  for (int i = 0; i < 3; ++i) {
+    if (memory[i].start == memory[i].end)
+      continue;
+    assert(memory[i].start < memory[i].end);
+    vm_physseg_plug_used(memory[i].start, memory[i].end);
+    if (i == 2) {
+      if (memory[i].end < ram_end)
+        vm_physseg_plug(memory[i].end, ram_end);
+    } else if (memory[i].end < memory[i + 1].start) {
+      vm_physseg_plug(memory[i].end, memory[i + 1].start);
```

Comment:
Could you explain it, despite that `memory[i+1]` (i+1 = 3) is not set above and it could be random?

Reply:
You have two cases:
- i == 2; then you want to add an area from the last occupied byte to the end of RAM (if it's possible)
- i < 2; then you want to add an area from the end of occupied region to the start of new region (if it's possible)

Would it be more readable if I changed 3 to a constant?

Reply:
My bad I misread the place of else statement, everything is fine.